File: entrega-21-09/typescript/bases.ts
/**
 * Subclasses concretas de Produto (HERANÇA): cada tecnologia de base tem um
 * atributo próprio e seu próprio peso de penalidade quando o acabamento não
 * bate com o perfil desejado. Todas reaproveitam `calcularScoreBase` da
 * superclasse e implementam `calcularCompatibilidade` de forma independente
 * (POLIMORFISMO — ver main.ts, onde o mesmo laço chama o método sem saber
 * qual subclasse está por trás de cada `Produto`).
 */

import { Produto, PerfilDesejado, Subtom, Acabamento } from "./produto";

function arredondar(valor: number): number {
  return Math.round(valor * 10) / 10;
}

/** Base líquida: penalidade leve de acabamento (tecnologia mais versátil). */
export class BaseLiquida extends Produto {
  static readonly PESO_PENALIDADE_ACABAMENTO = 5;

  #fps: number = 0;

  constructor(
    marca: string,
    nome: string,
    preco: number,
    tom: string,
    subtom: Subtom,
    acabamento: Acabamento,
    fps: number
  ) {
    super(marca, nome, preco, tom, subtom, acabamento);
    this.fps = fps;
  }

  get fps(): number {
    return this.#fps;
  }

  set fps(valor: number) {
    if (valor < 0) {
      throw new RangeError("fps nao pode ser negativo");
    }
    this.#fps = valor;
  }

  calcularCompatibilidade(perfilDesejado: PerfilDesejado): number {
    const scoreBase = this.calcularScoreBase(perfilDesejado);
    const penalidade =
      this.acabamento !== perfilDesejado.acabamento
        ? BaseLiquida.PESO_PENALIDADE_ACABAMENTO
        : 0;
    const scoreFinal = Math.max(0, scoreBase - penalidade);
    return arredondar(scoreFinal);
  }
}

/** Base cushion: penalidade alta de acabamento (tecnologia mais rígida). */
export class BaseCushion extends Produto {
  static readonly PESO_PENALIDADE_ACABAMENTO = 15;

  #numeroEsponjas: number = 1;

  constructor(
    marca: string,
    nome: string,
    preco: number,
    tom: string,
    subtom: Subtom,
    acabamento: Acabamento,
    numeroEsponjas: number
  ) {
    super(marca, nome, preco, tom, subtom, acabamento);
    this.numeroEsponjas = numeroEsponjas;
  }

  get numeroEsponjas(): number {
    return this.#numeroEsponjas;
  }

  set numeroEsponjas(valor: number) {
    if (valor <= 0) {
      throw new RangeError("numeroEsponjas deve ser maior que zero");
    }
    this.#numeroEsponjas = valor;
  }

  calcularCompatibilidade(perfilDesejado: PerfilDesejado): number {
    const scoreBase = this.calcularScoreBase(perfilDesejado);
    const penalidade =
      this.acabamento !== perfilDesejado.acabamento
        ? BaseCushion.PESO_PENALIDADE_ACABAMENTO
        : 0;
    const scoreFinal = Math.max(0, scoreBase - penalidade);
    return arredondar(scoreFinal);
  }
}

/** Base mousse: penalidade intermediária de acabamento. */
export class BaseMousse extends Produto {
  static readonly PESO_PENALIDADE_ACABAMENTO = 10;

  #percentualAeracao: number = 0;

  constructor(
    marca: string,
    nome: string,
    preco: number,
    tom: string,
    subtom: Subtom,
    acabamento: Acabamento,
    percentualAeracao: number
  ) {
    super(marca, nome, preco, tom, subtom, acabamento);
    this.percentualAeracao = percentualAeracao;
  }

  get percentualAeracao(): number {
    return this.#percentualAeracao;
  }

  set percentualAeracao(valor: number) {
    if (valor < 0 || valor > 100) {
      throw new RangeError("percentualAeracao deve estar entre 0 e 100");
    }
    this.#percentualAeracao = valor;
  }

  calcularCompatibilidade(perfilDesejado: PerfilDesejado): number {
    const scoreBase = this.calcularScoreBase(perfilDesejado);
    const penalidade =
      this.acabamento !== perfilDesejado.acabamento
        ? BaseMousse.PESO_PENALIDADE_ACABAMENTO
        : 0;
    const scoreFinal = Math.max(0, scoreBase - penalidade);
    return arredondar(scoreFinal);
  }
}

File: entrega-21-09/python/bases.py
"""
bases.py
--------
Subclasses concretas de Produto (Herança + Polimorfismo).

Cada subclasse representa um formato de base diferente, com um
atributo próprio (também encapsulado) e um peso de penalidade de
acabamento diferente aplicado sobre o score base herdado de Produto.
Isso faz com que a mesma chamada polimórfica
`produto.calcular_compatibilidade(perfil)` produza resultados
diferentes conforme o tipo concreto do objeto, mesmo com a mesma
lógica de score base.
"""

from __future__ import annotations

from produto import PerfilDesejado, Produto, Subtom, Acabamento


class BaseLiquida(Produto):
    """Base líquida tradicional. Possui FPS (fator de proteção solar)."""

    PESO_PENALIDADE_ACABAMENTO = 5

    def __init__(
        self,
        marca: str,
        nome: str,
        preco: float,
        tom: str,
        subtom: Subtom,
        acabamento: Acabamento,
        fps: int,
    ) -> None:
        super().__init__(marca, nome, preco, tom, subtom, acabamento)
        self.fps = fps  # passa pelo setter, que valida

    @property
    def fps(self) -> int:
        return self._fps

    @fps.setter
    def fps(self, valor: int) -> None:
        if valor < 0:
            raise ValueError("fps não pode ser negativo")
        self._fps = valor

    def calcular_compatibilidade(self, perfil_desejado: PerfilDesejado) -> float:
        score_base = self._calcular_score_base(perfil_desejado)
        penalidade = (
            self.PESO_PENALIDADE_ACABAMENTO
            if self.acabamento != perfil_desejado.acabamento
            else 0
        )
        score_final = max(0.0, score_base - penalidade)
        return round(score_final, 1)


class BaseCushion(Produto):
    """Base cushion, aplicada com esponja embutida na embalagem."""

    PESO_PENALIDADE_ACABAMENTO = 15

    def __init__(
        self,
        marca: str,
        nome: str,
        preco: float,
        tom: str,
        subtom: Subtom,
        acabamento: Acabamento,
        numero_esponjas: int,
    ) -> None:
        super().__init__(marca, nome, preco, tom, subtom, acabamento)
        self.numero_esponjas = numero_esponjas  # passa pelo setter, que valida

    @property
    def numero_esponjas(self) -> int:
        return self._numero_esponjas

    @numero_esponjas.setter
    def numero_esponjas(self, valor: int) -> None:
        if valor <= 0:
            raise ValueError("numero_esponjas deve ser maior que zero")
        self._numero_esponjas = valor

    def calcular_compatibilidade(self, perfil_desejado: PerfilDesejado) -> float:
        score_base = self._calcular_score_base(perfil_desejado)
        penalidade = (
            self.PESO_PENALIDADE_ACABAMENTO
            if self.acabamento != perfil_desejado.acabamento
            else 0
        )
        score_final = max(0.0, score_base - penalidade)
        return round(score_final, 1)


class BaseMousse(Produto):
    """Base em textura mousse, com percentual de aeração da fórmula."""

    PESO_PENALIDADE_ACABAMENTO = 10

    def __init__(
        self,
        marca: str,
        nome: str,
        preco: float,
        tom: str,
        subtom: Subtom,
        acabamento: Acabamento,
        percentual_aeracao: float,
    ) -> None:
        super().__init__(marca, nome, preco, tom, subtom, acabamento)
        self.percentual_aeracao = percentual_aeracao  # passa pelo setter, que valida

    @property
    def percentual_aeracao(self) -> float:
        return self._percentual_aeracao

    @percentual_aeracao.setter
    def percentual_aeracao(self, valor: float) -> None:
        if not (0 <= valor <= 100):
            raise ValueError("percentual_aeracao deve estar entre 0 e 100")
        self._percentual_aeracao = valor

    def calcular_compatibilidade(self, perfil_desejado: PerfilDesejado) -> float:
        score_base = self._calcular_score_base(perfil_desejado)
        penalidade = (
            self.PESO_PENALIDADE_ACABAMENTO
            if self.acabamento != perfil_desejado.acabamento
            else 0
        )
        score_final = max(0.0, score_base - penalidade)
        return round(score_final, 1)
